temporian/implementation/numpy_cc/operators/window.cc

```diff
@@ -10,436 +10,566 @@
-        {
-          return;
-        }
+  void Remove(INPUT value) override {
+    if constexpr (std::numeric_limits<INPUT>::has_quiet_NaN) {
+      if (std::isnan(value)) {
+        return;
       }
-      sum_values -= value;
     }
+    sum_values -= value;
+  }
 
-    OUTPUT Result() override { return sum_values; }
-
-    // Sum of the values in the rolling window (RW).
-    double sum_values = 0;
-  };
-
-  template <typename INPUT, typename OUTPUT>
-  struct MovingExtremumAccumulator : Accumulator<INPUT, OUTPUT>
-  {
-    virtual ~MovingExtremumAccumulator() = default;
-    virtual bool Compare(INPUT a, INPUT b) = 0;
-
-    void Add(INPUT value) override
-    {
-      if constexpr (std::numeric_limits<INPUT>::has_quiet_NaN)
-      {
-        if (std::isnan(value))
-        {
-          return;
-        }
-      }
+  OUTPUT Result() override { return sum_values; }
+
+  // Sum of the values in the rolling window (RW).
+  double sum_values = 0;
+};
+
+template <typename INPUT, typename OUTPUT>
+struct MovingExtremumAccumulator : Accumulator<INPUT, OUTPUT> {
+  virtual ~MovingExtremumAccumulator() = default;
+  virtual bool Compare(INPUT a, INPUT b) = 0;
 
-      if (values.empty() || Compare(value, current_extremum))
-      {
-        // The value is the new
-        current_extremum = value;
+  void Add(INPUT value) override {
+    if constexpr (std::numeric_limits<INPUT>::has_quiet_NaN) {
+      if (std::isnan(value)) {
+        return;
       }
-      values.push_back(value);
     }
+    if (values.empty() || Compare(value, current_extremum)) {
+      current_extremum = value;
+    }
+    values.push_back(value);
+  }
 
-    void Remove(INPUT value) override
-    {
-      if constexpr (std::numeric_limits<INPUT>::has_quiet_NaN)
-      {
-        if (std::isnan(value))
-        {
-          return;
-        }
+  void AddLeft(INPUT value) override {
+    if constexpr (std::numeric_limits<INPUT>::has_quiet_NaN) {
+      if (std::isnan(value)) {
+        return;
       }
+    }
+    if (values.empty() || Compare(value, current_extremum)) {
+      current_extremum = value;
+    }
+    values.push_front(value);
+  }
 
-      if (values.size() == 1)
-      {
-        values.clear();
+  void Remove(INPUT value) override {
+    if constexpr (std::numeric_limits<INPUT>::has_quiet_NaN) {
+      if (std::isnan(value)) {
+        return;
       }
-      else
-      {
-        assert(!values.empty());
-        assert(values.front() == value);
-        values.pop_front();
-        if (value == current_extremum)
-        {
-          // Compute the extremum on the remaining items.
-          current_extremum = values.front();
-          for (const auto value : values)
-          {
-            if (Compare(value, current_extremum))
-            {
-              current_extremum = value;
-            }
+    }
+
+    assert(values.front() == value);
+    if (values.size() == 1) {
+      values.clear();
+    } else {
+      assert(!values.empty());

```

Comment:
Nit: I would add an assert that, before the pop, the front value of "values" is equal to "value".

Reply:
assert is a couple lines above